### PR TITLE
Adds Unified APIs SentryTransport

### PR DESCRIPTION
**Pull request by fetzig**
* `SentryTransport` has been extracted from `SentryClient`
 * To fetch config `SentryOption` is used as single point of truth (no more `SentryClient.enabled` but `SentryClient.options.enabled`, a.s.o.)

### CODE REVIEW

File: Sources/Sentry/SentryTransport.m [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+}
+
+
+- (void)setMaxEvents:(NSUInteger)maxEvents {
+    self.fileManager.maxEvents = maxEvents;
+}
+
+- (void)setMaxBreadcrumbs:(NSUInteger)maxBreadcrumbs {
+    self.fileManager.maxBreadcrumbs = maxBreadcrumbs;
+}
```
Comment (HazAT) [inline Sources/Sentry/SentryTransport.m]: We can completely remove this.

File: Sources/Sentry/SentryTransport.m [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+- (void)setSampleRate:(float)sampleRate {
+    if (sampleRate < 0 || sampleRate > 1) {
+        [SentryLog logWithMessage:@"sampleRate must be between 0.0 and 1.0" andLevel:kSentryLogLevelError];
+        return;
+    }
+    _sampleRate = sampleRate;
+    self.shouldSendEvent = ^BOOL(SentryEvent *_Nonnull event) {
+        return (sampleRate >= ((double)arc4random() / 0x100000000));
+    };
+}
```
Comment (HazAT) [inline Sources/Sentry/SentryTransport.m]: Please move this logic into the client, the transport should really just store/send events.

File: Sources/Sentry/include/SentryClient.h [hunk 3]
```diff
@@ -45,77 +47,20 @@ SENTRY_NO_INIT
...
- * This property will be filled before the event is sent.
- */
-@property(nonatomic, copy) NSString *_Nullable releaseName;
 
 /**
- * This property will be filled before the event is sent.
+ * This block can be used to modify the event before it will be serialized and sent
  */
-@property(nonatomic, copy) NSString *_Nullable dist;
+@property(nonatomic, copy) SentryBeforeSerializeEvent _Nullable beforeSerializeEvent;
```
Comment (HazAT) [inline Sources/Sentry/include/SentryClient.h]: Note, this should be a property on the Client anymore, instead this should become the `beforeSend` callback in the options, we can do this in another PR.

File: Sources/Sentry/include/SentryClient.h [hunk 4]
```diff
@@ -45,77 +47,20 @@ SENTRY_NO_INIT
...
-/**
- * Contains the last successfully sent event
- */
-@property(nonatomic, strong) SentryEvent *_Nullable lastEvent;
-    
-/**
- * Is the client enabled?. Default is @YES, if set @NO sending of events will be prevented.
- */
-@property(nonatomic, copy) NSNumber *enabled;
+@property(nonatomic, strong) SentryTransport* transport;
```
Comment (HazAT) [inline Sources/Sentry/include/SentryClient.h]: Property shouldn't be public, people should just be able to set it through the options.

File: Sources/Sentry/include/SentryClient.h [hunk 5]
```diff
@@ -137,38 +82,11 @@ SENTRY_NO_INIT
...
- * This function is mainly used for react native.
- * @param event SentryEvent that should be sent
- */
-- (void)storeEvent:(SentryEvent *)event scope:(SentryScope *)scope;
-
-/**
- * Clears all context related variables: tags, extra and user
- */
-- (void)clearContext;
+- (void)captureEvent:(SentryEvent *)event withScope:(SentryScope *)scope;
```
Comment (HazAT) [inline Sources/Sentry/include/SentryClient.h]: `scope` should be nullable.

File: Sources/Sentry/include/SentryClient.h (hunk 4 above)
Comment (fetzig) [reply Sources/Sentry/include/SentryClient.h]: true. But what to do with `SentryBeforeSendRequest`? I would put it in options if that makes sense to you.

File: Sources/Sentry/SentryTransport.m (hunk 1 above)
Comment (fetzig) [reply Sources/Sentry/SentryTransport.m]: done

File: Sources/Sentry/include/SentryClient.h (hunk 3 above)
Comment (fetzig) [reply Sources/Sentry/include/SentryClient.h]: done

File: Sources/Sentry/include/SentryClient.h (hunk 5 above)
Comment (fetzig) [reply Sources/Sentry/include/SentryClient.h]: done

File: Sources/Sentry/SentryTransport.m (hunk 2 above)
Comment (fetzig) [reply Sources/Sentry/SentryTransport.m]: done

File: Sources/Sentry/include/SentryClient.h (hunk 4 above)
Comment (HazAT) [reply Sources/Sentry/include/SentryClient.h]: Yes, put it in the options and just call it `SentryBeforeSend`

File: Sources/Sentry/include/SentryClient.h [hunk 6]
```diff
@@ -47,19 +47,23 @@ SENTRY_NO_INIT
...
  * when sending the crashreport
  */
 @property(nonatomic, strong) NSDictionary<NSString *, id> *_Nullable lastContext;
 
-@property(nonatomic, strong) SentryTransport* transport;
+/**
+ * This block can be used to prevent the event from being sent.
+ * @return BOOL
+ */
+@property(nonatomic, copy) SentryShouldSendEvent _Nullable shouldSendEvent;
```
Comment (HazAT) [inline Sources/Sentry/include/SentryClient.h]: We don't need this anymore, if `beforeSend` returns `nil` it should not send the event, we probably need to do the sample rate differently and not in this callback. More in `prepareEvent`
